### PR TITLE
Persistent docker session

### PR DESCRIPTION
**Pull request by SmartManoj**
Solves #1811

Set PERSIST_SESSION=1 and SSH_PORT, SSH_PASSWORD, CONTAINER_ID from the existing docker sandbox container.

### CODE REVIEW

File: opendevin/runtime/docker/ssh_box.py [hunk 1]
```diff
@@ -232,28 +232,32 @@ def __init__(
...
+            config.container_id or self.container_name_prefix + self.instance_id
+        )
 
         # set up random user password
-        self._ssh_password = str(uuid.uuid4())
-        self._ssh_port = find_available_tcp_port()
-
+        self._ssh_password = config.ssh_password or str(uuid.uuid4())
+        self._ssh_port = config.ssh_port or find_available_tcp_port()
         # always restart the container, cuz the initial be regarded as a new session
```
Comment (rbren) [inline opendevin/runtime/docker/ssh_box.py:242]:
```suggestion
```

File: opendevin/core/config.py [hunk 2]
```diff
@@ -173,6 +173,10 @@ class AppConfig(metaclass=Singleton):
     disable_color: bool = False
     sandbox_user_id: int = os.getuid() if hasattr(os, 'getuid') else 1000
     sandbox_timeout: int = 120
+    persist_session: bool = False
+    ssh_port: int | None = None
+    ssh_password: str | None = None
+    container_id: str | None = None
```
Comment (rbren) [inline opendevin/core/config.py:179]: we should probably factor out a SandboxConfig

**Comment by rbren** [comment]
This is really neat!

I wonder if there's a reasonable way to make this the default behavior. We'd have to pick a port and hard-code a secret...

File: opendevin/core/config.py (hunk 2 above)
Comment (enyst) [reply opendevin/core/config.py:179]: Definitely. It doesn't have to be this PR though IMHO.

**Comment by SmartManoj** [comment]
https://github.com/OpenDevin/OpenDevin/pull/847#pullrequestreview-1984874251

File: opendevin/core/utils/singleton.py [hunk 3]
```diff
@@ -22,6 +22,9 @@ def reset(cls):
         for instance_type, instance in cls._instances.items():
             print('resetting... ', instance_type)
             for field in dataclasses.fields(instance_type):
+                if field.name == 'persist_sandbox':
+                    # do not reset the persist_sandbox flag
+                    continue
```
Comment (enyst) [inline opendevin/core/utils/singleton.py:27]: Just curious, why did you need this? I think this method is unused except for a few pytest tests. Or was unused

File: opendevin/runtime/docker/ssh_box.py [hunk 4]
```diff
@@ -199,7 +199,7 @@ class DockerSSHBox(Sandbox):
     def __init__(
         self,
         container_image: str | None = None,
-        timeout: int = 120,
+        timeout: int = 180,
```
Comment (enyst) [inline opendevin/runtime/docker/ssh_box.py:202]: Isn't this the timeout after which a background command is killed? The default of 120 seems quite high since the app looks hung. 😔 But maybe I'm not seeing this right, I'm looking on the web interface for the moment

Comment (enyst) [reply opendevin/runtime/docker/ssh_box.py:202]: Exactly. The default timeout shows up in issues like https://github.com/OpenDevin/OpenDevin/issues/1280 and not only for interactive programs, any command that the agent runs, which doesn't return fast. The app looks stuck from the user perspective, for the duration.

That said, I thought this init is called normally with a timeout, set by config.sandbox_timeout, so the default here won't make a difference. So I'm curious why you set it higher.

File: opendevin/core/utils/singleton.py (hunk 3 above)
Comment (SmartManoj) [reply opendevin/core/utils/singleton.py:27]: image

Comment (enyst) [reply opendevin/core/utils/singleton.py:27]: Yes, and if it is called from tests, it should reset everything. They're tests, at the end of the each test, everything should be clean, unchanged, like default.

That said... I think the reset() method should not exist at all, we should remove it some time. Tests can be cleaned up without it.

Comment (enyst) [inline opendevin/core/utils/singleton.py:27]:
```suggestion
```

Comment (SmartManoj) [reply opendevin/core/utils/singleton.py:27]: But current tests require clean sandboxes

Comment (enyst) [reply opendevin/core/utils/singleton.py:27]: Ah, I see what happens. I made a PR to remove these resets from the sandboxes tests. It works both on local and on CI, so it really should not be there at all.

Then the env var you set in this PR should be respected.

File: opendevin/core/config.py [hunk 5]
```diff
@@ -179,6 +179,10 @@ class AppConfig(metaclass=Singleton):
     disable_color: bool = False
     sandbox_user_id: int = os.getuid() if hasattr(os, 'getuid') else 1000
     sandbox_timeout: int = 120
+    persist_sandbox: bool = True
+    ssh_port: int = 63710
+    # https://xkcd.com/936/
+    ssh_password: str = 'correct horse battery staple1'
```
[the remark this answers is not in the document]
Comment (SmartManoj) [reply opendevin/core/config.py:185]: https://devops.stackexchange.com/a/3904

[the remark this answers is not in the document]
Comment (SmartManoj) [reply opendevin/core/config.py:185]: https://github.com/OpenDevin/OpenDevin/pull/891/files#r1616428972 This will solve?

[the remark this answers is not in the document]
Comment (rbren) [reply opendevin/core/config.py:185]: No, IMO we need to make sure there aren't any hard-coded default passwords. Some users won't know that they need to change them, and will be open to vulnerabilities

File: README.md [hunk 6]
```diff
@@ -54,17 +54,19 @@ To start the app, run these commands, replacing `$(pwd)/workspace` with the dire
...
 ```
 
-> [!WARNING]  
-> OpenDevin runs bash commands within a Docker sandbox, so it should not affect your machine. 
+> [!WARNING]
+> OpenDevin runs bash commands within a Docker sandbox, so it should not affect your machine.
 > But your workspace directory will be attached to that sandbox, and files in the directory may be modified or deleted.
 
 ```bash
 docker run \
```
Comment (rbren) [inline README.md:65]: There's another `docker run` command in the docs folder too--we'll need to change there as well

**Review by rbren — APPROVED** [review]
Thanks @SmartManoj!

File: agenthub/codeact_agent/codeact_agent.py [hunk 7]
```diff
@@ -196,7 +196,7 @@ def step(self, state: State) -> Action:
             {'role': 'system', 'content': self.system_message},
             {
                 'role': 'user',
-                'content': f"Here is an example of how you can interact with the environment for task solving:\n{EXAMPLES}\n\nNOW, LET'S START!",
+                'content': f"Here is an example of how you can interact with the environment for task solving:\n{EXAMPLES}\n\nNOW, LET'S START!\n",
```
Comment (li-boxuan) [inline agenthub/codeact_agent/codeact_agent.py:199]: Is there any reason for this prompt change?

Comment (SmartManoj) [reply agenthub/codeact_agent/codeact_agent.py:199]: The given task was appended to that without any spaces. Now it is separated by message_separator. It is not needed now.

Comment (li-boxuan) [reply agenthub/codeact_agent/codeact_agent.py:199]: Oh I see. I reverted this when resolving conflicts for https://github.com/OpenDevin/OpenDevin/pull/2105 - benchmarks for that version have already been run so we don't want to introduce anything that might change the prompt. Feel free to leave a comment on the next version of CodeActAgent. It should come soon...

File: opendevin/runtime/docker/ssh_box.py [hunk 8]
```diff
@@ -216,38 +216,48 @@ def __init__(
...
-                )
-                n_tries -= 1
-                if n_tries == 0:
-                    raise e
-                time.sleep(5)
-        self.setup_user()
-
+        if config.persist_sandbox:
+            if not config.ssh_password:
+                raise Exception('Password must be set for persistent sandbox')
```
Comment (xingyaoww) [inline opendevin/runtime/docker/ssh_box.py:231]: Maybe we could make the error message more informative! Maybe like "please add `ssh_password` to your `config.toml` or add `-e SSH_PASSWORD` to your `docker run` command"

Comment (SmartManoj) [reply opendevin/runtime/docker/ssh_box.py:231]: Done in #2139

File: opendevin/core/config.py (hunk 2 above)
Comment (SmartManoj) [reply opendevin/core/config.py]: It should be done at that time. Now one needs to update the configs.
